### PR TITLE
Implement resampling with mean

### PR DESCRIPTION
**Pull request by achow**
@jeeyoungk @Nathan-Fenner

### CODE REVIEW

**Comment by jeeyoungk** [comment]
lol, this feels very similar to https://github.com/square/metrics/pull/60

File: api/backend/blueflood/blueflood.go [hunk 1]
```diff
@@ -198,8 +198,56 @@ func bluefloodResolution(r int64) string {
...
+	series := make([]float64, timerange.Slots())
+
+	// Cursor into `series`
+	idx := 0
+	// Cursors into `points` for working data windows
+	windowStart, windowEnd := 0, 0
+	// Generate points at timestamps corresponding to the timerange
+	for t := timerange.Start(); t <= timerange.End(); t, idx = t+timerange.Resolution(), idx+1 {
+		// Set up our series window
+		for ; points[windowStart].Timestamp < t; windowStart++ {
```
Comment (Nathan-Fenner) [inline api/backend/blueflood/blueflood.go]: `windowStart` can overrun the end of `points` and cause a panic.

Comment (jeeyoungk) [inline api/backend/blueflood/blueflood.go]: i think here the assumption is `points[...].Timestamp` is between `[Start(), End()]`

it's a bit subtle, maybe add a condition for `windowStart < Slots`?

File: api/backend/blueflood/blueflood.go [hunk 2]
```diff
@@ -198,8 +198,56 @@ func bluefloodResolution(r int64) string {
...
+		for windowEnd = windowStart; windowEnd < len(points)-1 && points[windowEnd+1].Timestamp < (t+timerange.Resolution()); windowEnd++ {
+		}
+
+		switch sampleMethod {
+		case api.SampleMean:
+			// Uses weighted average over [t, t + resolution)
+			totalPoints, runningTotal := 0, 0.0
+			for i := windowStart; i <= windowEnd; i++ {
+				totalPoints += points[i].Points
+				runningTotal += points[i].Average * float64(points[i].Points)
```
Comment (jeeyoungk) [inline api/backend/blueflood/blueflood.go]: i dislike the implicit assumption to `Average` is weird. maybe use nathan's solution to extract fields?

Comment (achow) [inline api/backend/blueflood/blueflood.go]: Resampling using mean over (for example) max values is meaningless, imo.

**Pull request closed without merging**